050-Kafka-Node-Pools.md

```diff
@@ -0,0 +1,806 @@
+    strimzi.io/cluster: my-cluster
+spec:
+  role:
+    - broker
+  replicas: 6
+  resources:
+    requests:
+      memory: 64Gi
+      cpu: 8
+    limits:
+      memory: 64Gi
+      cpu: 12
+  storage:
+    type: jbod
+    volumes:
+    - id: 0
+      type: persistent-claim
+      size: 1000Gi
+    - id: 1
+      type: persistent-claim
+      size: 1000Gi
+```
+
+### Handling of node IDs
+
+Node IDs (broker IDs) uniquely identify each of the Kafka cluster nodes.
+Node IDs will be assigned sequentially:
+
+* They will start with 0.
+* By default, when a new node ID is needed, the lowest possible available number will be used regardless the pool where the node will be added:
+    * In most cases, the node ID will be assigned from 0 up: 0, 1, 2, 3.
+    * If any gap will exist in the existing node IDs, it will be filled in.
+      For example when the existing nodes have IDs 0, 1, 4, 5, 6; the next node will get ID 2.
+* By default, when scaling down a pool, a highest used node ID will be remove.
+    * For example in Kafka cluster with two pool, where the first pool has IDs 0, 1 and 2 and the second pool has IDs 3, 4 and 5 ... when you scale down the first pool, the node with ID 2 will be removed and when you scale down the second pool node 5 will be removed.
+* The pods will reflect the node IDs in their name in the same way as they do today.
+
+The fact that the node IDs will not be assigned to the pools means the Pods will be mixed.
+In a cluster with two pools, they might look like this:
+
+* `my-cluster-big-nodes-0`
+* `my-cluster-big-nodes-1`
+* `my-cluster-small-nodes-2`
+* `my-cluster-small-nodes-3`
+* `my-cluster-big-nodes-4`
+* `my-cluster-small-nodes-5`
+
+This is anyway unavoidable if we want to allow nodes to move between the pools.
+
+To allow the operator to keep track of the assigned node IDs, it will keep track of them in the `.status` section of the `KafkaNodePool` resource.
+That way, it will be able to know at any point in time which IDs are used and which are available.
+It will be also able to identify the scale-down or scale-up events using this information.
+
+```yaml=
+apiVersion: kafka.strimzi.io/v1beta2
+kind: KafkaNodePool
+metadata:
+  name: big-nodes
+  # ...
+spec:
+  replicas: 3
+  # ...
+status:
+  nodeIds:
+    - 0
+    - 1
+    - 2
+  clusterId: cABpUKeDTqi7q_XOtBMgUv
+  # replicas and labelSelector fields are needed for the scale subresource
+  replicas: 3
+  labelSelector: >-
+    strimzi.io/cluster=my-cluster,strimzi.io/name=m-cluster-big-nodes,strimzi.io/kind=Kafka
+```
+
+User will be able to use annotations to indicate to the operator what should be the next ID when scaling up or what should be the node ID which should be removed when scaling down through annotations.
+Setting an annotation `strimzi.io/next-node-ids` will tell the operator what should be the next ID used in scale-up.
+The value might be a single number, list of numbers or one or more ranges.
+E.g. `[3]`, `[3, 4, 5]` or `[1000-1010]`.
```

Comment:
Assuming the user have a node pool with 3 brokers (so ids 0,1,2) and `strimzi.io/next-node-ids: [3]`.
He scales up from 3 to 4 through the `replicas` field so operator will create broker id 3.
Then with same `strimzi.io/next-node-ids` annotation, the user will scale up again ... now ... because 3 is used, will the operator raise a warning about it but using id = 4 or failing the reconcile asking the user to fix the annotation? I think that not having consistency between current max id and the annotation value can drive to confusion.

Reply:
This is not part of my prototype - at least not yet. But the annotations are expected to be treated as an indication of what should be used next. So if we cannot fulfill them, we would pick up the next available ID and use it. This goes better with the ideas of autoscaling etc. It also IMHO better fits the concept of a list where some might be taken and some not etc.

Reply:
> So if we cannot fulfill them, we would pick up the next available ID and use it

That's fine with me. At least just printing a WARNING like "ID 3 is already in use, picking the next available one 4"